src/devices/machine/spifi3.h

```diff
@@ -0,0 +1,247 @@
+	uint32_t intr_r();
+	void intr_w(uint32_t data);
+	uint32_t imask_r();
+	void imask_w(uint32_t data);
+	uint32_t prstat_r();
+	uint32_t init_status_r();
+	uint32_t fifoctrl_r();
+	void fifoctrl_w(uint32_t data);
+	uint32_t data_xfer_r();
+	void data_xfer_w(uint32_t data);
+	uint32_t autocmd_r();
+	void autocmd_w(uint32_t data);
+	uint32_t autostat_r();
+	void autostat_w(uint32_t data);
+	uint32_t select_r();
+	void select_w(uint32_t data);
+	void prcmd_w(uint32_t data);
+	uint32_t auxctrl_r();
+	void auxctrl_w(uint32_t data);
+	uint32_t autodata_r();
+	void autodata_w(uint32_t data);
+	uint32_t identify_r();
+	void identify_w(uint32_t data);
+	uint32_t scsi_status_r();
+	void scsi_status_w(uint32_t data);
+	uint32_t icond_r();
+	void icond_w(uint32_t data);
+	uint32_t fastwide_r();
+	void fastwide_w(uint32_t data);
+	uint32_t exctrl_r();
+	void exctrl_w(uint32_t data);
+
+	// Command buffer accessors
+	uint8_t cmd_buf_r(offs_t offset);
+	void cmd_buf_w(offs_t offset, uint8_t data);
+
+	struct spifi_cmd_entry
+	{
+		// NetBSD has these mapped as uint32_t to align the accesses and such
+		// in reality, these are all 8-bit values that are mapped, in typical NWS-5000 series
+		// fashion, to be 32-bit word aligned.
+		// the same probably applies to the register file.
+		uint8_t cdb[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+		uint8_t quecode = 0;
+		uint8_t quetag = 0;
+		uint8_t idmsg = 0;
+		uint8_t status = 0;
+	};
+
+	struct register_file
+	{
+		uint32_t spstat = 0;
+		uint32_t cmlen = 0;
+		uint32_t cmdpage = 0;
+		// count_hi, count_mid, count_low
+
+		uint32_t svptr_hi = 0;
+		uint32_t svptr_mid = 0;
+		uint32_t svptr_low = 0;
+
+		uint32_t intr = 0;
+		uint32_t imask = 0;
+		uint32_t prctrl = 0;
+
+		uint32_t prstat = 0;
+		uint32_t init_status = 0;
+		uint32_t fifoctrl = 0;
+		uint32_t fifodata = 0;
+
+		uint32_t config = 0;
+		uint32_t data_xfer = 0;
+		uint32_t autocmd = 0;
+		uint32_t autostat = 0;
+
+		uint32_t resel = 0;
+		uint32_t select = 0;
+		// prcmd, which is used to trigger commands
+		uint32_t auxctrl = 0;
+
+		uint32_t autodata = 0;
+		uint32_t loopctrl = 0;
+		uint32_t loopdata = 0;
+		uint32_t identify = 0;
+
+		uint32_t complete = 0;
+		uint32_t scsi_status = 0x1; // Must be 0x1 for SPIFI to be recognized at boot
+		uint32_t data = 0;
+		uint32_t icond = 0;
+
+		uint32_t fastwide = 0;
+		uint32_t exctrl = 0;
+		uint32_t exstat = 0;
+		uint32_t test = 0;
+
+		uint32_t quematch = 0;
+		uint32_t quecode = 0;
+		uint32_t quetag = 0;
+		uint32_t quepage = 0;
+
+		spifi_cmd_entry cmbuf[8];
+	} spifi_reg;
+
+	// State tracking variables
+	dma_direction dma_dir;
+	scsi_mode mode;
+	scsi_data_target xfr_data_source;
+	int state;
+	int xfr_phase;
+	int command_pos;
+	bool irq = false;
+	bool drq = false;
+	uint32_t tcounter;
+	uint8_t sync_period;
+	uint8_t clock_conv = 2; // TODO: appropriate value for SPIFI
+	emu_timer *tm;
+	int bus_id;
+	std::queue<uint32_t> m_even_fifo;
+	std::queue<uint32_t> m_odd_fifo;
```

Comment:
The `int` and `std::queue` embers are going to cause issues for save state support.

Reply:
Sure, I'll go ahead and change those. Is there any library of save-state-safe data structures like queues and such? The closest thing I was able to find was `util::fifo` in `coretmpl.h`. Just want to make sure I'm not reinventing the wheel before implementing the queue with an array or whatever.

Reply:
There isn’t a library of save state safe structures yet, unfortunately.  You kind of do have to reinvent the wheel for things that are too small to justify being a device in their own right, but fortunately a lot of them are relatively simple.